### PR TITLE
Improvements for the KnowledgeBase and some metrics for Processes.

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/tycho/kb/KnowledgeBase.java [hunk 1]
```diff
@@ -247,18 +247,20 @@ public List<KnowledgeBaseArticle> query(String language, String query, int maxRe
...
-               .queryString(query, QueryField.contains(KnowledgeBaseEntry.SEARCH_FIELD))
+               .eqIgnoreNull(KnowledgeBaseEntry.LANG, language)
+               .queryString(query,
+                            QueryField.contains(KnowledgeBaseEntry.SEARCH_FIELD))
                .limit(maxResults * 5)
                .iterateAll(entry -> {
                    if (entry.checkPermissions()) {
                        result.add(new KnowledgeBaseArticle(entry, language, this));
                    }
                });
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/kb/KnowledgeBase.java:258]: Optional: extract Method for query and permission check which calls consumer if entry can be accessed (could replace both very similar queries)

File: src/main/java/sirius/biz/tycho/kb/KnowledgeBaseController.java [hunk 2]
```diff
@@ -116,4 +119,21 @@ private boolean checkArticlePermission(KnowledgeBaseArticle article, String auth
...
     }
+
+    /**
+     * Provides an autocompletion for KB articles.
+     *
+     * @param webContext the current request
+     */
+    @Permission(TenantUserManager.PERMISSION_SYSTEM_TENANT_MEMBER)
+    @Routed(value = "/kb/autocomplete", priority = 99)
+    public void tenantsAutocomplete(final WebContext webContext) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/kb/KnowledgeBaseController.java:130]:
```suggestion
    public void articlesAutocomplete(final WebContext webContext) {
```

File: src/main/java/sirius/biz/tycho/kb/KnowledgeBaseController.java [hunk 3]
```diff
@@ -116,4 +119,21 @@ private boolean checkArticlePermission(KnowledgeBaseArticle article, String auth
...
+        AutocompleteHelper.handle(webContext, (query, result) -> {
+            knowledgeBase.query(null, query, 25).forEach(article -> {
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/kb/KnowledgeBaseController.java:132]: Optional: Define a `AutocompleteHelper#DEFAULT_LIMIT` constant (also note OX uses a limit of 10 in most places, maybe we should unify that)

File: src/main/java/sirius/biz/tycho/kb/KnowledgeBaseEntry.java [hunk 4]
```diff
@@ -108,6 +109,11 @@ public class KnowledgeBaseEntry extends SearchableEntity {
     public static final Mapping SYNC_ID = Mapping.named("syncId");
     private String syncId;
 
+    @BeforeSave
+    protected void onSave() {
+        setSearchableContent(getArticleId());
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/kb/KnowledgeBaseEntry.java:114]: Cant use the Annotation on the field instead?

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

File: src/main/java/sirius/biz/tycho/kb/KnowledgeBaseController.java (hunk 3 above)
Comment (andyHa) [reply src/main/java/sirius/biz/tycho/kb/KnowledgeBaseController.java:132]: https://github.com/scireum/sirius-web/pull/1092

**Review by sabieber — APPROVED** [review]

**Review by cschierle — APPROVED** [review]

**Pull request merged**